modules/repository-s3/src/test/java/org/elasticsearch/repositories/s3/CustomWebIdentityTokenCredentialsProviderTests.java

```diff
@@ -124,8 +157,64 @@ public void testCreateWebIdentityTokenCredentialsProvider() throws Exception {
                 webIdentityTokenCredentialsProvider
             ).getCredentials();
 
-            Assert.assertEquals("sts_access_key", credentials.getAWSAccessKeyId());
-            Assert.assertEquals("secret_access_key", credentials.getAWSSecretKey());
+            assertCredentials(credentials);
+        } finally {
+            webIdentityTokenCredentialsProvider.shutdown();
+            httpServer.stop(0);
+        }
+    }
+
+    private static class DelegatingConsumer implements Consumer<String> {
+        private Consumer<String> delegate;
+
+        private DelegatingConsumer(Consumer<String> delegate) {
+            this.delegate = delegate;
+        }
+
+        private void setDelegate(Consumer<String> delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public void accept(String s) {
+            delegate.accept(s);
+        }
+    }
+
+    @SuppressForbidden(reason = "HTTP server is used for testing")
+    public void testPickUpNewWebIdentityTokenWhenItsChanged() throws Exception {
+        DelegatingConsumer webIdentityTokenCheck = new DelegatingConsumer(s -> assertEquals("YXdzLXdlYi1pZGVudGl0eS10b2tlbi1maWxl", s));
+
+        HttpServer httpServer = getHttpServer(webIdentityTokenCheck);
+        Environment environment = getEnvironment();
+        Map<String, String> environmentVariables = environmentVariables();
+        Map<String, String> systemProperties = getSystemProperties(httpServer);
+        var webIdentityTokenCredentialsProvider = new S3Service.CustomWebIdentityTokenCredentialsProvider(
+            environment,
+            environmentVariables::get,
+            systemProperties::getOrDefault,
+            Clock.fixed(Instant.ofEpochMilli(1651084775908L), ZoneOffset.UTC),
+            resourceWatcherService
+        );
+        try {
+            AWSCredentialsProvider awsCredentialsProvider = S3Service.buildCredentials(
+                LogManager.getLogger(S3Service.class),
+                S3ClientSettings.getClientSettings(Settings.EMPTY, randomAlphaOfLength(8)),
+                webIdentityTokenCredentialsProvider
+            );
+            assertCredentials(awsCredentialsProvider.getCredentials());
+
+            String newWebIdentityToken = "88f84342080d4671a511e10ae905b2b0";
+            Files.writeString(environment.configFile().resolve("repository-s3/aws-web-identity-token-file"), newWebIdentityToken);
+
+            var latch = new CountDownLatch(1);
+            webIdentityTokenCheck.setDelegate(s -> {
+                if (s.equals(newWebIdentityToken)) {
+                    latch.countDown();
+                }
+            });
```

Comment:
This needs to be placed before updating the token file. There is a chance that the refresh process gets triggered and completes before this block of code can run.